src/physics.cpp

```diff
@@ -275,130 +283,135 @@ void sample_photon_reaction(Particle& p)
     p.u() = rotate_angle(p.u(), mu, nullptr, p.current_seed());
     p.event_ = TallyEvent::SCATTER;
     p.event_mt_ = COHERENT;
-    return;
-  }
-
-  // Incoherent (Compton) scattering
-  prob += micro.incoherent;
-  if (prob > cutoff) {
-    double alpha_out, mu;
-    int i_shell;
-    element.compton_scatter(alpha, true, &alpha_out, &mu, &i_shell, p.current_seed());
-
-    // Determine binding energy of shell. The binding energy is 0.0 if
-    // doppler broadening is not used.
-    double e_b;
-    if (i_shell == -1) {
-      e_b = 0.0;
-    } else {
-      e_b = element.binding_energy_[i_shell];
-    }
-
-    // Create Compton electron
-    double phi = 2.0*PI*prn(p.current_seed());
-    double E_electron = (alpha - alpha_out)*MASS_ELECTRON_EV - e_b;
-    int electron = static_cast<int>(Particle::Type::electron);
-    if (E_electron >= settings::energy_cutoff[electron]) {
-      double mu_electron = (alpha - alpha_out*mu)
-        / std::sqrt(alpha*alpha + alpha_out*alpha_out - 2.0*alpha*alpha_out*mu);
-      Direction u = rotate_angle(p.u(), mu_electron, &phi, p.current_seed());
-      p.create_secondary(u, E_electron, Particle::Type::electron);
-    }
-
-    // TODO: Compton subshell data does not match atomic relaxation data
-    // Allow electrons to fill orbital and produce auger electrons
-    // and fluorescent photons
-    if (i_shell >= 0) {
-      const auto& shell = element.shells_[i_shell];
-      element.atomic_relaxation(shell, p);
-    }
-
-    phi += PI;
-    p.E_ = alpha_out*MASS_ELECTRON_EV;
-    p.u() = rotate_angle(p.u(), mu, &phi, p.current_seed());
-    p.event_ = TallyEvent::SCATTER;
-    p.event_mt_ = INCOHERENT;
-    return;
-  }
+  } else { // add by Yuan
```

Comment:
Hrmmm I can see why this section of the code has been restructured to avoid some of the `return` statements before getting to the `split_particle` call. To avoid these more deeply nested if-else statements, we can call our new `split_particle` function inside of `collision`. (Same for the neutron splitting to make it consistent.)

```cpp
switch (p.type_) {
  case Particle::Type::neutron:
    sample_neutron_reaction(p);
    if (settings::weightwindow && settings::n_ww)  split_particle(p);
    break;
  case Particle::Type::photon:
    sample_photon_reaction(p);
    if (settings::weightwindow && settings::p_ww)  split_particle(p);
    break;
  case Particle::Type::electron:
    sample_electron_reaction(p);
    break;
  case Particle::Type::positron:
    sample_positron_reaction(p);
    break;
  }
```

Then we can restore the code in `sample_photon_reaction`.

Reply:
Yes, that will be more clear.